subarray.go

```diff
@@ -0,0 +1,263 @@
+		ep := reflect.New(reflect.ValueOf(r.end).Type())
+		ep.Elem().Set(reflect.ValueOf(r.end))
+		ret = C.tiledb_subarray_add_range_by_name(sa.context.tiledbContext, sa.subarray, cDimName,
+			sp.UnsafePointer(), ep.UnsafePointer(), nil)
+	}
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error adding subarray range: %s", sa.context.LastError())
+	}
+
+	return nil
+}
+
+// GetRangeNum retrieves the number of ranges of the query subarray along a given dimension index.
+func (sa *Subarray) GetRangeNum(dimIdx uint32) (uint64, error) {
+	var rangeNum uint64
+
+	ret := C.tiledb_subarray_get_range_num(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx), (*C.uint64_t)(unsafe.Pointer(&rangeNum)))
+	if ret != C.TILEDB_OK {
+		return 0, fmt.Errorf("Error retrieving subarray range num: %s", sa.context.LastError())
+	}
+
+	return rangeNum, nil
+}
+
+// GetRangeNum retrieves the number of ranges of the query subarray along a given dimension name.
+func (sa *Subarray) GetRangeNumFromName(dimName string) (uint64, error) {
+	var rangeNum uint64
+
+	cDimName := C.CString(dimName)
+	defer C.free(unsafe.Pointer(cDimName))
+
+	ret := C.tiledb_subarray_get_range_num_from_name(sa.context.tiledbContext, sa.subarray, cDimName, (*C.uint64_t)(unsafe.Pointer(&rangeNum)))
+	if ret != C.TILEDB_OK {
+		return 0, fmt.Errorf("Error retrieving subarray range num: %s", sa.context.LastError())
+	}
+
+	return rangeNum, nil
+}
+
+// GetRange retrieves a specific range of the subarray along a given dimension index.
+func (sa *Subarray) GetRange(dimIdx uint32, rangeNum uint64) (Range, error) {
+	dt, isVar, err := datatypeOfDimensionFromIndex(sa.array, dimIdx)
+	if err != nil {
+		return Range{}, err
+	}
+
+	var r Range
+	var ret C.int32_t
+	if isVar {
+		var startSize, endSize uint64
+		ret = C.tiledb_subarray_get_range_var_size(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx), C.uint64_t(rangeNum),
+			(*C.uint64_t)(unsafe.Pointer(&startSize)), (*C.uint64_t)(unsafe.Pointer(&endSize)))
+		if ret == C.TILEDB_OK {
+			var sp, ep unsafe.Pointer
+			var startData, endData []byte
+			if startSize > 0 {
+				startData = make([]byte, int(startSize))
+				sp = unsafe.Pointer(&startData[0])
+			}
+			if endSize > 0 {
+				endData = make([]byte, int(endSize))
+				ep = unsafe.Pointer(&endData[0])
+			}
+			ret = C.tiledb_subarray_get_range_var(sa.context.tiledbContext, sa.subarray,
+				C.uint32_t(dimIdx), C.uint64_t(rangeNum), sp, ep)
+			if ret == C.TILEDB_OK {
+				r.start = string(startData)
+				r.end = string(endData)
+			}
+		}
+	} else {
+		var startPointer, endPointer, stridePointer unsafe.Pointer
+		ret = C.tiledb_subarray_get_range(sa.context.tiledbContext, sa.subarray,
+			C.uint32_t(dimIdx), C.uint64_t(rangeNum), &startPointer, &endPointer, &stridePointer)
+		typ := dt.ReflectType()
+		if ret == C.TILEDB_OK {
+			r.start = reflect.NewAt(typ, startPointer).Elem().Interface()
+			r.end = reflect.NewAt(typ, endPointer).Elem().Interface()
```

Comment:
It looks to me like this is going to give us a pointer into unmanaged memory. What guarantees do we have over the lifetime of the memory that is passed back to us? How do we ensure that it neither leaks nor is prematurely freed?

Reply:
The pointer is in memory managed by TileDB core (unsafe as Go is concerned) but i use it only temporarily to copy the value.

```
                var startPointer, endPointer, stridePointer unsafe.Pointer
                ret = C.tiledb_subarray_get_range(sa.context.tiledbContext, sa.subarray,
                        C.uint32_t(dimIdx), C.uint64_t(rangeNum), &startPointer, &endPointer, &stridePointer)
		typ := dt.ReflectType()
                if ret == C.TILEDB_OK {
                        r.start = reflect.NewAt(typ, startPointer).Elem().Interface()
                        r.end = reflect.NewAt(typ, endPointer).Elem().Interface()
                }
```

I use the `startPointer etc` to get the range from TileDB and then i copy the value of that pointer reflectively (via `Elem`) The value is copied to Range and the pointers are not longer used, so no leaks

Reply:
Understood. Thank you for explaining!